src/qilib/data_set/memory_data_set_io.py

```diff
@@ -0,0 +1,153 @@
+"""Quantum Inspire library
+
+Copyright 2019 QuTech Delft
+
+qilib is available under the [MIT open-source license](https://opensource.org/licenses/MIT):
+
+Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated
+documentation files (the "Software"), to deal in the Software without restriction, including without limitation the
+rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+persons to whom the Software is furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+"""
+
+import queue
+from collections import deque
+from typing import Any, Tuple, Union, Optional
+
+from qilib.data_set import DataSetIO, DataArray, DataSet
+
+
+class MemoryDataSetIO(DataSetIO):
+    """ An interface that allows synchronization between different instances of a DataSet in the same thread."""
+
+    def __init__(self) -> None:
+        """ The MemoryDataSetIO contains DataArrays and metadata of measurements.
+        
+            The interface is used for synchronization between different instances of a DataSet in the same thread.
+            The functionality is made for implementing live plotting of a DataSet.
+        """
+        self.__user_data_storage: Any = queue.Queue()
+        self.__data_array_storage: Any = queue.Queue()
+        self._is_finalized = False
+        self._data_set: DataSet
```

Comment:
Why not give it a value, like None, and then ask in `__is_bounded` if it is None or not? instead of defining an instance attribute outside of `__init__`? (which upsets static analysis)

Reply:
I reverted to `self._data_set = None`. Initially I changed this because MyPy was complaining. Looking at it more closely, I think this will be automatically resolved when data_set is type-hinted.